contracts/core/EntryPoint.sol

```diff
@@ -581,42 +585,80 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
      * @param opIndex                            - The operation index.
      * @param op                                 - The user operation.
      * @param opInfo                             - The operation info.
-     * @param requiredPreFund                    - The required prefund amount.
      * @return context                           - The Paymaster-provided value to be passed to the 'postOp' function later
      * @return validationData                    - The Paymaster's validationData.
      */
     function _validatePaymasterPrepayment(
         uint256 opIndex,
         PackedUserOperation calldata op,
-        UserOpInfo memory opInfo,
-        uint256 requiredPreFund
+        UserOpInfo memory opInfo
     ) internal virtual returns (bytes memory context, uint256 validationData) {
         unchecked {
             uint256 preGas = gasleft();
             MemoryUserOp memory mUserOp = opInfo.mUserOp;
             address paymaster = mUserOp.paymaster;
+            uint256 requiredPreFund = opInfo.prefund;
             if (!_tryDecrementDeposit(paymaster, requiredPreFund)) {
                 revert FailedOp(opIndex, "AA31 paymaster deposit too low");
             }
             uint256 pmVerificationGasLimit = mUserOp.paymasterVerificationGasLimit;
-            try
-                IPaymaster(paymaster).validatePaymasterUserOp{gas: pmVerificationGasLimit}(
-                    op,
-                    opInfo.userOpHash,
-                    requiredPreFund
-                )
-            returns (bytes memory _context, uint256 _validationData) {
-                context = _context;
-                validationData = _validationData;
-            } catch {
-                revert FailedOpWithRevert(opIndex, "AA33 reverted", Exec.getReturnData(REVERT_REASON_MAX_LEN));
-            }
+            (context, validationData) = _callValidatePaymasterUserOp(opIndex, op, opInfo);
             if (preGas - gasleft() > pmVerificationGasLimit) {
                 revert FailedOp(opIndex, "AA36 over paymasterVerificationGasLimit");
             }
         }
     }
 
+    function _callValidatePaymasterUserOp(
+        uint256 opIndex,
+        PackedUserOperation calldata op,
+        UserOpInfo memory opInfo
+    ) internal returns (bytes memory context, uint256 validationData)  {
+        uint256 freePtr = _getFreePtr();
+        bytes memory validatePaymasterCall = abi.encodeCall(
+            IPaymaster.validatePaymasterUserOp,
+            (op, opInfo.userOpHash, opInfo.prefund)
+        );
+        address paymaster = opInfo.mUserOp.paymaster;
+        bool success;
+        uint256 contextLength;
+        uint256 contextOffset;
+        uint256 maxContextLength;
+        assembly ("memory-safe") {
+            //call and return 3 first words: offset, validation, context-length
+            success := call(gas(), paymaster, 0, add(validatePaymasterCall, 0x20), mload(validatePaymasterCall), freePtr, 96)
```

Comment:
I think that removing `verificationGasLimit` and `paymasterVerificationGasLimit` from the call here is not good: we introduce another unpredictable factor to simulation vs execution of this validation, depending on the transaction gas limit provided by the bundler to `handleOps()`, and it's almost always going to differ between simulating a userOp and simulating the entire bundle. 
Even if we don't see the issue now, is there a benefit of introducing this?

Reply:
I added it here, since that was the behaviour beforehand.
but still: I think we should remove the verification-limit from teh calls, since the actual value is checked with a `preGas-gasleft()`. that is, the gas-limit on the calls can never be reached, since the gas-limit should cover also calling into the function (copying data in reading result, and for verification-gas, it also covers creation and nonce increment.

moreover: currently, you can't tell OOG from revert.... but if we separate, you have a clear separation between AA36 (gas limit) and AA33 (revert) ...

it requires few tests to fix, because they chekc the wrong error code.